allennlp/semparse/worlds/quarel_world.py

```diff
@@ -0,0 +1,182 @@
+
+    @overrides
+    def get_valid_starting_types(self) -> Set[Type]:
+        return self.types.starting_types
+
+    # Simple table for how attributes relates to each other
+    # First entry is by convention (above in __init__) the friction subset
+    qr_coeff_sets_default = [
+            {"friction": 1, "speed": -1, "smoothness": -1, "distance": -1, "heat": 1},
+            {"speed": 1, "time": -1},
+            {"speed": 1, "distance": 1},
+            {"time": 1, "distance": 1},
+            {"weight": 1, "acceleration": -1},
+            {"strength": 1, "distance": 1},
+            {"strength": 1, "thickness": 1},
+            {"mass": 1, "gravity": 1},
+            {"flexibility": 1, "breakability": -1},
+            {"distance": 1, "loudness": -1, "brightness": -1, "apparentSize": -1},
+            {"exerciseIntensity": 1, "amountSweat": 1}
+    ]
+
+    # Size translation for absolute and relative values
+    qr_size = {
+            'higher': 1,
+            'high': 1,
+            'low': -1,
+            'lower': -1
+    }
+
+    def _get_qr_coeff(self, attr1, attr2):
+        for qset in self.qr_coeff_sets:
+            if attr1 in qset and attr2 in qset:
+                return qset[attr1] * qset[attr2]
+        return 0
+
+    def _check_compatible(self, setup: List, answer: List) -> bool:
+        attributes = {setup[0], answer[0]}
+        qr_coeff = None
+        for qr_coeff_set in self.qr_coeff_sets:
+            if not attributes - qr_coeff_set.keys():
+                qr_coeff = qr_coeff_set
+        if qr_coeff is None:
+            return False  # No compatible attribute sets found
+
+        attribute_dir = qr_coeff[setup[0]] * qr_coeff[answer[0]]
+        change_same = 1 if self.qr_size[setup[1]] == self.qr_size[answer[1]] else -1
+        world_same = 1 if setup[2] == answer[2] else -1
+        return attribute_dir * change_same * world_same == 1
+
+    def _exec_infer(self, setup, *answers):
+        answer_index = -1
+        if len(answers) == 1:
+            if self._check_compatible(setup, answers[0]):
+                return 1
+            else:
+                return 0
+        for index, answer in enumerate(answers):
+            if self._check_compatible(setup, answer):
+                if answer_index > -1:
+                    # Found two valid answers
+                    answer_index = -2
+                else:
+                    answer_index = index
+        return answer_index
+
+    def _exec_and(self, expr):
+        if not expr or expr[0] != 'and':
+            return expr
+        args = expr[1:]
+        if len(args) == 1:
+            return args[0]
+        if len(args) > 2:
+            # More than 2 arguments not allowed by current grammar
+            return None
+        if self._check_compatible(args[0], args[1]):
+            # Check that arguments are compatible, then fine to keep just one
+            return args[0]
+        return None
+
+    def execute(self, lf_raw: str) -> int:
```

Comment:
If this gets any more complicated, we'll probably want to move the execution logic into `executors`.